src/libraries/System.Linq/src/System/Linq/First.cs

```diff
@@ -31,12 +31,21 @@ public static TSource First<TSource>(this IEnumerable<TSource> source, Func<TSou
         }
 
         public static TSource? FirstOrDefault<TSource>(this IEnumerable<TSource> source) =>
-            source.TryGetFirst(out bool _);
+            source.TryGetFirst(out _);
+
+        public static TSource FirstOrDefault<TSource>(this IEnumerable<TSource> source, TSource defaultValue) =>
+            source.TryGetFirst(defaultValue, out _)!;
 
         public static TSource? FirstOrDefault<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate) =>
-            source.TryGetFirst(predicate, out bool _);
+            source.TryGetFirst(predicate, out _);
+
+        public static TSource FirstOrDefault<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate, TSource defaultValue) =>
+            source.TryGetFirst(predicate, defaultValue, out _)!;
 
-        private static TSource? TryGetFirst<TSource>(this IEnumerable<TSource> source, out bool found)
+        private static TSource? TryGetFirst<TSource>(this IEnumerable<TSource> source, out bool found) =>
+            source.TryGetFirst(default(TSource), out found);
+
+        private static TSource? TryGetFirst<TSource>(this IEnumerable<TSource> source, TSource defaultValue, out bool found)
```

Comment:
I think we could minimize the amount of nullability assertions if we reworked the signature of this and similar methods to like so:
```suggestion
        private static bool TryGetFirst<TSource>(this IEnumerable<TSource> source, [MaybeNullWhen(false)] out TSource result)
```

It should then be relatively straightforward to express the public methods that depend on it like so:
```csharp
        public static TSource? FirstOrDefault<TSource>(this IEnumerable<TSource> source)
        {
            source.TryGetFirst(out TSource? first);
            return first;
        }

        public static TSource FirstOrDefault<TSource>(this IEnumerable<TSource> source, TSource defaultValue) =>
            source.TryGetFirst(out TSource? result) ? result : defaultValue;
```

And remove the need for stub methods like the one in line 45.

@stephentoub are there any performance considerations that necessitate keeping the current shape of `out` parameters? Only thing I could think of is TCO but to my knowledge C# does not emit tailcall opcodes.

Reply:
> are there any performance considerations that necessitate keeping the current shape of out parameters?

If TSource is a reference type, returning it via the `out` will incur a write barrier.  Returning it as a return value won't.  You can see that here:
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunHgIIAGwYBBABQB5YACsYYDAwCWAOwAOAVwxoGETQymz5OzRowBKDlx7XdGUwwC8StZoDcV692IB2BhijqMO7U1gC+HgwR9DoycgoAQpKxRiqm2rYM/EIMuOpgYDC4uBacIZ7cufmFuI5+AUER1j7OpsFhNKFAA===
```C#
using System;
public class C {
    public bool A(Object input, out Object output) {
        output = input;
        return true;
    }
    
    public object B(Object input, out bool success) {
        success = true;
        return input;
    }
}
```
producing
```
C.A(System.Object, System.Object ByRef)
    L0000: mov rcx, r8
    L0003: call 0x00007ffefcb69f50
    L0008: mov eax, 1
    L000d: ret

C.B(System.Object, Boolean ByRef)
    L0000: mov byte ptr [r8], 1
    L0004: mov rax, rdx
    L0007: ret
```
That `call 0x00007ffefcb69f50` is the overhead.

Reply:
Thanks, in that case we should just keep it as-is.